src/main/groovy/wooga/gradle/build/unity/tasks/AbstractUnityBuildEngineTask.groovy

```diff
@@ -0,0 +1,131 @@
+package wooga.gradle.build.unity.tasks
+
+
+import org.gradle.api.file.RegularFile
+import org.gradle.api.file.RegularFileProperty
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.provider.Property
+import org.gradle.api.provider.Provider
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputFile
+import org.gradle.api.tasks.Optional
+import wooga.gradle.build.unity.internal.BuildEngineArgs
+import wooga.gradle.secrets.internal.Secrets
+import wooga.gradle.unity.UnityTask
+
+import javax.crypto.spec.SecretKeySpec
+
+
+class AbstractUnityBuildEngineTask extends UnityTask {
+
+    private final Property<String> build
+    private final Property<String> exportMethodName
+    private final Property<String> outputPath
+    private final RegularFileProperty secretsFile
+    private final Property<SecretKeySpec> secretsKey
+    private final ListProperty<Object> customArguments
+
+    AbstractUnityBuildEngineTask() {
+        this.build = project.objects.property(String)
+        this.exportMethodName = project.objects.property(String)
+        this.outputPath = project.objects.property(String)
+        this.secretsFile = project.objects.fileProperty()
+        this.secretsKey = project.objects.property(SecretKeySpec.class)
+        this.customArguments = project.objects.listProperty(Object)
+    }
+
+    BuildEngineArgs defaultArgs() {
+        Provider<File> outputDir = outputPath.map { path ->
+            def pathFile = new File(path)
+            return pathFile.isAbsolute() ? pathFile : new File(project.projectDir, path)
+        }
+        def secrets = secretsFile.map { RegularFile secretsFile ->
+            Secrets.decode(secretsFile.asFile.text)
+        }
+        def environmentSecrets = project.provider({
+            generateSecretsEnvironment(secrets, secretsKey)
+        }.memoize())
+
+        def buildEngineArgs = new BuildEngineArgs(project.providers, exportMethodName)
+        buildEngineArgs.with {
+            addArgs(customArguments.orElse(new ArrayList<>()))
+            addArg("--build", build)
+            addArg("--outputPath", outputDir.map{out -> out.path})
+            addEnvs(environmentSecrets)
+        }
+        return buildEngineArgs
+    }
+
+    def setupExecution(BuildEngineArgs unityArgs) {
+        environment.putAll(unityArgs.environment)
+        additionalArguments.add(unityArgs.customArgsStr)
+        additionalArguments.add("-executeMethod")
+        additionalArguments.add(unityArgs.method)
+    }

```

Comment:
I'm not a huge fan of this construct and the need to call super at a specific time.

Reply:
It does not need to be called an specific time, it can be called anytime, as long as a `BuildEngineArgs` is provided.